### PR TITLE
[TEST] Add VI Overclock

### PR DESCRIPTION
**Pull request by Sam-Belliveau**
image
image

**AUDIO SPEED IS NOT AFFECTED BY VI OVERCLOCK SETTINGS**

This is a much more reliable way of inducing a frame skip, and likely will induce much fewer bugs than VI Skip. It's main advantage is being able to increase the framerate, however this will 100% affect game speed when physics is tied to framerate.

I'm going to mark this as draft because it is definitely not usable at the moment.

### CODE REVIEW

**Comment by JMC47** [comment]
I'm going to have to be skeptical about this.  Most games with frameskip and whatnot won't really go *above* 60 FPS even with hacks in the past.

**Comment by mbc07** [comment]
Regarding the UI, current design of this PR makes the Advanced tab even taller than it already is and further exacerbates the problem PR #8785 tries to address. With this PR we'll have enough UI controls related to overclocking to justify a new, dedicated tab, for them...

**Comment by SuperSamus** [comment]
If someone wants to toy around with this today, I rebased it [here](https://github.com/SuperSamus/dolphin/tree/vi-overclock).
(Note that I didn't bother solving Android-related merge conflicts)

**Comment by Calinou** [comment]
> If someone wants to toy around with this today, I rebased it [here](https://github.com/SuperSamus/dolphin/tree/vi-overclock). (Note that I didn't bother solving Android-related merge conflicts)

I gave this branch a try and it works well to push Star Fox Adventures to 120 FPS (although it ends up triggering the same bug as https://github.com/RenaKunisaki/StarFoxAdventures/issues/9). At least there are no more audio issues, unlike the previous approach where I increased emulation speed and decreased internal game speed.

**Comment by SuperSamus** [comment]
I briefly tested it on Skylanders: Spyro's Adventure with the [60 FPS patch](https://wiki.dolphin-emu.org/index.php?title=Skylanders:_Spyro%27s_Adventure#60_FPS)[^1].

The game runs at the same gameplay speed from 33% to 200% (it slows-down/speeds-up below/above that, it seems intentional[^2]), though running it at 120 FPS is **insanely** demanding (the game is pretty demanding as is...).

Here's some footage, half speed because my computer is too weak[^3]:

https://github.com/dolphin-emu/dolphin/assets/40663462/b019279e-638b-4913-ac8c-761913ad6195

[^1]: The name is a misnomer, the game already runs at 60 FPS normally, but renders everything except the UI at 30 FPS. The patch makes the game render at the same speed it's running, so it's an efficiency gain.
[^2]: Why would the devs intentionally make sure that the game speeds up above 120 FPS _on a Wii game_ is beyond me. [Even the PC version of the game is capped at 60 FPS](https://www.pcgamingwiki.com/wiki/Skylanders:_Spyro%27s_Adventure#Video), if PCGW isn't lying (do note that the not-Wii/WiiU versions are a different (presumably older) build with some weird differences). Anyway, I couldn't find the part of the code that does this...
[^3]: If you have déjà vu, it's because [I already posted it on Discord](https://discord.com/channels/521709831132807179/521711358794792961/1235325599271092274).

**Comment by SuperSamus** [comment]
So, experimented a little more with Skylanders: Spyro's Adventure, to go even further beyond.
<details>
<summary>Click if you are curious on the technical details</summary>

First, I found where the game was limiting the lower delta time (see previous post, note 2).
```
0409fe4c 60000000
```
Note:
- Low delta time = high FPS
- High delta time = low FPS


Actually maintaining these high framerates however, is wack: sometimes the game decides to not run at the framerate I want. The FPS could be close to the VPS, I raise the emulated CPU clock a little, and suddenly the FPS goes down to exactly half the VPS (the specific percentages depend on the game's situation). What is going on?

With the branch watch tool, I found the function called every vblank that makes the physics step and renders: 0x80171b9c.
This function first checks whether 0x9026f6fc is not 0: if true, it proceeds and reduces it by 1, otherwise it does nothing.
So, this variable has only one other setter: the function 0x80171a90, which simply increases it by 1, and does other mysterious stuff.
One weird thing is that, depending on the level of the game (I think?), 0x80171a90 is called one time every frame, or two times every two frames (the latter makes the problem worse, though it also exists in the former).
So, when the game is running at half the VPS, it's simply because 0x80171a90 isn't called as much as it should be. Who calls 0x80171a90?
Bad news: it seems to be something very low level, as it's called through a variable pointer (that also calls a lot of other stuff), and it's surrounded by scary OS syscalls.
And it's not easy to debug either: when pausing the debugger, the behavior is different from only logging the breakpoints.
So, I can't know why it's not called when it should be. Is it a problem with the way VI Overclock is implemented, or is it the game? (Maybe it's the game: see later)
The best I could do is remove the check from 0x80171b9c. While it does make the game do a physics step and render when it should, it also makes it do it when it shouldn't.
```
04171bb8 60000000
```
It's a small side effect though, since it's only visual (you'll see older frames *after* newer ones), and it can be prevented by overclocking the emulated CPU so that it never/rarely happens. The game looks perfectly stable otherwise, and performs as it should (note that the performance metrics will always say that the FPS are equal to the VPS, which is "true" only if the emulated CPU clock is high enough!).

Later, I noticed later that the lower/upper bound for delta times weren't "fixed", but depended on another address (which *is* fixed at 1/60): 0x8084f230.
It looks like a "target framerate" address? It's also read in one other place: 0x8000b5b0.
*Very* curiously, the latter uses it to affect the behavior I described above: if I lower the value, the VI Overclock will have to be higher for the game to start artificially limiting the framerate.
All of this, without the visual problems if the emulated CPU can't keep up.
There's a catch: you can't use ridiculously low values, or even 0. If you try, the game will crash after loading screens.
I could test out more values, but the former solution is satisfying enough to me. Feel free to experiment.
Example (sets to 1/120):
```
0484f230 3c088889
```
If you are testing with ridiculously low values, remember that the upper delta time limit will also be ridiculously low, leading to the game going in slow-motion. Dirty way to disable:
```
0409fe5c 60000000
```
</details>

TL;DR, If you want to play this game at higher than 120 FPS, use this patch, preferably along with the "[60 FPS patch](https://wiki.dolphin-emu.org/index.php?title=Skylanders:_Spyro%27s_Adventure#60_FPS)", and set the emulated CPU clock to max (otherwise you'll see old frames when you shouldn't).
```
$Even further beyond 120 FPS
04171bb8 60000000 # Force physics/render
0409fe4c 60000000 # Remove lower delta bound
```

And don't use VBI Overclock.

Also, what kind of computer do you need to actually run the game at 120 FPS in full speed? Mine is barely enough for 60.

**Comment by SuperSamus** [comment]
Here are the codes for Skylanders: Giants.
Same rules of the other one.
The 60 FPS patch is [here](https://wiki.dolphin-emu.org/index.php?title=Skylanders:_Giants#60_FPS).
Use this:
```
$Change target framerate
22701EEC 0040E8C7 # This address is used for something else during initialization
04701EEC 3C088889 # 1/120, replace with 1/VPS set in VI Overclock
E0000000 80008000
*If the value doesn't match 1/VPS, emulation performance will be worse, and the game may refuse to run at the framerate you want. To calculate hex, use https://www.h-schmidt.net/FloatConverter/IEEE754.html
```

Then, optionally:
```
$Delta time bounds fix
040AD180 48000009
040AD184 3D4CCCCD # Hardcode 1/20
040AD188 7C6802A6
040AD18C C0030000 # Load to f0
*Removes lower limit (by writing over it), and forces upper limit to 1/20
```

<details>
<summary>Alternative stuff (not recommended)</summary>

```
$Force physics/render step
0417E8C8 60000000 # Remove check
*If emulated CPU clock isn't high enough, you will see older frames. FPS will (falsely) always match VPS.
```
</details>

**Comment by SuperSamus** [comment]
Skylanders: Swap Force is also able to go to higher framerates, and it doesn't exhibit the weird behavior of the other two games.
There's a bug that softlocks you during sequences where there is a long fall[^1]. This issue also happens in all other versions of the game.
I found the cause thanks to branch watch, and patched it:
```
$Infinite fall fix
C2657438 00000003 # On top of `lfs f0, 0x005C (r29)`
3C004416 9001FFFC # lis r0, 0x4416 (20*30FPS); stw r0, -4(sp)
C001FFFC EC0007B2 # lfs f0, -4(sp); fmuls f0, f0, f30 # f30 is deltaTime
60000000 00000000
```
(A port of the fix for Cemu is coming soon™)

[^1]: After 3 seconds of falling, the game teleports you back to where the fall started, softlocking you in a loop. The reason is because of a failsafe meant in case your character is stuck falling between geometry. The game determines whether you are stuck with a simple method: it calculates the squared distance of your character from the position *of the previous frame*, then checks whether it's lower than 20². If it is true 3 seconds, then it teleports you. But if the framerate is higher, your distance is going to be lower than normal, and the game will consider you stuck even when falling at top speed. Ironic... The fix simply consists in changing the value from 20 to 20×30FPS×deltaTime.

**Comment by SuperSamus** [comment]
Gave it a try on Need For Speed: Nitro (R7XP69).
The game was sped up, but I made a code to adjust it:
```
$120 FPS
047763B4 3C088889
```

- The actual game works perfectly, but the menus are sped up (who cares).
- If you want to change the value to go even higher... you can't. Starting from somewhere between 1/126 and 1/130, the game just freezes while loading the race (regardless if your VI Overclock settings), complete with an invalid op popup. I tried setting it during the race, but again, the game freezes after a few seconds. I wonder if this is even fixable... Though, 120 FPS is still a nice upgrade.

Multiplayer works too, just use [this code](https://wiki.dolphin-emu.org/index.php?title=Need_for_Speed:_Nitro#60_FPS_in_Multiplayer) to prevent the game from capping to 30 FPS.

If you want to port it to other versions, just look for `3C888889` in the `.dol`, there are only a few of them, so you can easily do it by trial and error.

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 1]
```diff
@@ -292,6 +337,30 @@ void AdvancedPane::Update()
     return tr("%1% (%2 MHz)").arg(QString::number(percent), QString::number(clock));
   }());
 
+  QFont vi_bf = font();
+  vi_bf.setBold(Config::GetActiveLayerForConfig(Config::MAIN_VI_OVERCLOCK_ENABLE) !=
+                Config::LayerType::Base);
+  m_vi_clock_override_checkbox->setFont(bf);
```
Comment (Tilka) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:343]: `bf` was probably meant to be `vi_bf` here.

**Comment by SuperSamus** [comment]
Super Mario Sunshine at 120 FPS.

https://github.com/user-attachments/assets/5c053493-e3ad-4c03-bdb1-f435f6d5983d

Unfortunately, it's not really usable.
At 60 FPS (without VBI overclock), sometimes the music doesn't play, or gets stuck on the first note.
At 120 FPS, "sometimes" becomes "more often than not".

This is caused by the VBI frequency specifically. Lowering it before the music plays "solves" the problem (which is how I got the music for the video, it refused to play on the level otherwise). Apparently, [this is also hard to fix](https://github.com/DotKuribo/BetterSunshineEngine/issues/18). Too bad.

The code is a modified 60 FPS code (PAL) from the wiki. With the way this cheat works, 120 FPS is the maximum.
```
0440DD10 40000000
042F4CB4 60000000
0440BE54 3D23D70A
C20066EC 00000003 # I don't know what this does, so I'm not even sure if I should have changed it.
C2C28028 EC2105B2
EC2105B2 FEC00890
60000000 00000000
```

**Comment by SuperSamus** [comment]
Pikmin satisfies!

https://github.com/user-attachments/assets/36519694-5a14-446e-9c2b-26d0f71b1410

By default, the game doesn't speed up when changing the VBI overclock.

However, there are few (not game breaking) things that are wrong:
- ~~Cutscene speed~~ Fixed (the 60 FPS code on the wiki already does that, I just changed the value)
- ~~Pikmin give up going to an object too quickly~~ Fixed (by me)
- Onions spit up seeds too quickly
- Some particle effects are too fast, and generate too quickly
- Some animations play at a reduced framerate
- Some 2D animations are... too fancy (for instance, the ship in the area select rotates too much)

```
$120 FPS
04059774 38000001
0405AE7C 38800001
043E8A58 3E800000 # Cutscene speed
04107738 4800000C
040AFDEC 28000078 # Extend Pikmin "task give up" timer
```

```
$144 FPS
04059774 38000001
0405AE7C 38800001
043E8A58 3E555555 # Cutscene speed
04107738 4800000C
040AFDEC 28000090 # Extend Pikmin "task give up" timer
```

(Also, do note that normally, without cheat codes, some parts of the game (like the main menu) run at 60 FPS, while the main game runs at 30 FPS.)

**Comment by SuperSamus** [comment]
Crash Nitro Kart... just works perfectly.
No cheat codes, no nothing.

https://github.com/user-attachments/assets/38d4bb87-bbae-4632-aeca-a4feddff69fa

**Comment by SuperSamus** [comment]
Pikmin 2 (GC NTSC) 120 FPS code.
Same issues as the [60 FPS code](https://wiki.dolphin-emu.org/index.php?title=Pikmin_2_(GC)#60_FPS), but doubled.

```
$120 FPS
04423540 3BC00001 # `setFramerate`: force full FPS
04520430 42F00000 # 120 FPS
045204A8 42F00000 # Menu 120 FPS
0451D804 3FB00000 # Map translation speed 1
C2312B04 00000002 # Map translation speed 2
3C003F00 9001FFFC
C121FFFC 00000000
0451D7B0 3BF5C28F # Map scale speed
04520628 3F811111 # Cutscene delta time
0452062C 20000000
C2010654 00000002 # Quadruple cutscene sections' "frame duration"
5484103A 909F002C
60000000 00000000
C242ED28 00000002 # Cutscene: objects spawn/despawn timing
3C003E80 9001FFFC
C001FFFC 00000000
C24374DC 00000002 # Cutscene: trigger `startFadeblack` a frame earlier (fixes black screen in first cutscene)
806D9AEC C0230054
FC020840 00000000
```

Doing it for framerates that aren't $30*2^i$ is a little impractical, due to what's done with `C2010654`.

**Comment by SuperSamus** [comment]
Multi-game extravaganza, with games developed by Beenox with their Goliath Engine.
The games using the engine are:
- Bee Movie Game (at least the PS2/PC/X360 versions, I don't know if the Wii version uses the same engine)
- Monsters vs. Aliens
- Skylanders: SuperChargers Racing (tested)
- Spider-Man: Edge of Time (RPCS3 wiki says that vblank works until 60 FPS)
- Spider-Man: Shattered Dimensions (tested PC version, RPCS3 wiki says that vblank works until 60 FPS)
- The Amazing Spider-Man
- (Maybe more? There's barely any documentation...)

The games support dynamic frame rates, but have two FPS locks:
- A half-VPS lock (that is, they currently run at 30 FPS)[^1]
- A 60 FPS lock, though the way it's implemented makes it a little higher (about 62 FPS)[^2]

This means that by raising the VBI Overclock to 200%, these games will effortlessly run at 60 FPS.

But if you want to go higher, changing the second lock is super easy. It's just a 1/60 (`3C888889`) value in memory.

For instance, with this code on Skylanders: SuperChargers Racing (PAL):
```
046F2F60 00000000
```
 And 480% VBI Overclock, I can play at 144 FPS flawlessly.

[^1]: https://forums.dolphin-emu.org/Thread-anyone-interested-in-making-a-60fps-code-for-beenox-spiderman-games-hint-inside?pid=536725#pid536725
[^2]: On the PC version of Spider-Man: Shattered Dimensions, I found that some parts of the game are affected per-frame not by delta time, but by the value of the lock. For instance, if the actual FPS was 1/7th or lower than the lock, basic movement was completely borked. So... Maybe set the lock to the value you intend to play at, instead of `00000000`.

**Comment by JMC47** [comment]
Superseded by a newer PR.

**Comment by Calinou** [comment]
- Superseded by https://github.com/dolphin-emu/dolphin/pull/13046.
